scipy/ndimage/_filters.py

```diff
@@ -329,7 +332,8 @@ def gaussian_filter(input, sigma, order=0, output=None,
     >>> ax2.imshow(result)
     >>> plt.show()
     """
-    input = numpy.asarray(input)
+    xp, _ = get_namespace(input)
+    input = xp.asarray(input)
```

Comment:
This is great!

A code example verifying the behavior:
```py
In [1]: import numpy as np

In [2]: import numpy.array_api as npx
<ipython-input-2-f2c2051749b9>:1: UserWarning: The numpy.array_api submodule is still experimental. See NEP 47.
  import numpy.array_api as npx

In [3]: from scipy.ndimage import gaussian_filter

In [4]: a = npx.asarray(np.arange(50, step=2).reshape((5, 5)))

In [5]: type(a)
Out[5]: numpy.array_api._array_object.Array

In [6]: type(gaussian_filter(a, sigma=1))
Out[6]: numpy.array_api._array_object.Array
```

Reply:
Ahh, it doesn't work with CuPy! 😢 
```py
In [1]: import cupy as cp

In [2]: import cupy.array_api as cpx
<ipython-input-2-774ec64895d0>:1: UserWarning: The numpy.array_api submodule is still experimental. See NEP 47.
  import cupy.array_api as cpx

In [3]: from scipy.ndimage import gaussian_filter

In [4]: a = cpx.asarray(cp.arange(50, step=2).reshape((5, 5)))

In [5]: type(a)
Out[5]: cupy.array_api._array_object.Array

In [6]: type(gaussian_filter(a, sigma=1))
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [6], in <module>
----> 1 type(gaussian_filter(a, sigma=1))

File ~/.conda/envs/scipy-meson/lib/python3.10/site-packages/scipy/ndimage/_filters.py:346, in gaussian_filter(input, sigma, order, output, mode, cval, truncate)
    344 if len(axes) > 0:
    345     for axis, sigma, order, mode in axes:
--> 346         gaussian_filter1d(input, sigma, axis, order, output,
    347                           mode, cval, truncate)
    348         input = output
    349 else:

File ~/.conda/envs/scipy-meson/lib/python3.10/site-packages/scipy/ndimage/_filters.py:264, in gaussian_filter1d(input, sigma, axis, order, output, mode, cval, truncate)
    262 # Since we are calling correlate, not convolve, revert the kernel
    263 weights = _gaussian_kernel1d(sigma, order, lw)[::-1]
--> 264 return correlate1d(input, weights, axis, output, mode, cval, 0)

File ~/.conda/envs/scipy-meson/lib/python3.10/site-packages/scipy/ndimage/_filters.py:124, in correlate1d(input, weights, axis, output, mode, cval, origin)
    120     output = _ni_support._get_output(output, input, complex_output=True)
    121     return _complex_via_real_components(correlate1d, input, weights,
    122                                         output, cval, **kwargs)
--> 124 output = _ni_support._get_output(output, input)
    125 weights = numpy.asarray(weights, dtype=numpy.float64)
    126 if weights.ndim != 1 or weights.shape[0] < 1:

File ~/.conda/envs/scipy-meson/lib/python3.10/site-packages/scipy/ndimage/_ni_support.py:97, in _get_output(output, input, shape, complex_output)
     95     output = xp.zeros(shape, dtype=output)
     96 elif output.shape != shape:
---> 97     raise RuntimeError("output shape not correct")
     98 elif complex_output and output.dtype.kind != 'c':
     99     raise RuntimeError("output must have complex dtype")

RuntimeError: output shape not correct
```

Reply:
I see that in the demo `cupyx.scipy.ndimage.gaussian_filter` is used. In the future, when Anirudh's PR that adds dispatching to `scipy.ndimage` is merged, we would be able to pass CuPy arrays to the gaussian_filter and it would work nicely.
Here, let's use "if-else" dispatcher inside `scipy.ndimage.gaussian_filter`:
```py
# if xp is cupy.array_api:
return xp.asarray(
    cupyx.scipy.ndimage.gaussian_filter(
        input._array,
        sigma,
        order=order,
        output=output,
        mode=mode,
        cval=cval,
        truncate=truncate,
    )
)
```